aiida/engine/processes/workchains/restart.py

```diff
@@ -0,0 +1,425 @@
+                self.report('{}<{}> failed with an unrecoverable problem'.format(
+                    self.ctx.calc_name, calculation.pk))
+
+            # If the exit code returned actually has status `0` that means we consider the calculation as successful and
+            # set the corresponding flag such that the workchain will continue to the `results` step.
+            if isinstance(exit_code, ExitCode) and exit_code.status == 0:
+                self.ctx.is_finished = True
+
+        return exit_code
+
+    def results(self):  # pylint: disable=inconsistent-return-statements
+        """Attach the outputs specified in the output specification from the last completed calculation."""
+        calculation = self.ctx.calculations[self.ctx.iteration - 1]
+
+        # We check the `is_finished` attribute of the work chain and not the successfulness of the last calculation
+        # because the error handlers in the last iteration can have qualified a "failed" calculation as satisfactory
+        # for the outcome of the work chain and so have marked it as `is_finished=True`.
+        if not self.ctx.is_finished and self.ctx.iteration >= self.inputs.max_iterations.value:
+            self.report('reached the maximum number of iterations {}: last ran {}<{}>'.format(
+                self.inputs.max_iterations.value, self.ctx.calc_name, calculation.pk))
+            return self.exit_codes.ERROR_MAXIMUM_ITERATIONS_EXCEEDED  # pylint: disable=no-member
+
+        self.report('work chain completed after {} iterations'.format(self.ctx.iteration))
+
+        for name, port in self.spec().outputs.items():
+
+            try:
+                node = calculation.get_outgoing(link_label_filter=name).one().node
+            except ValueError:
+                if port.required:
+                    self.report("required output '{}' was not an output of {}<{}>".format(
+                        name, self.ctx.calc_name, calculation.pk))
+            else:
+                self.out(name, node)
+                if self._verbose:
+                    self.report("attaching the node {}<{}> as '{}'".format(node.__class__.__name__, node.pk, name))
+
+    def __init__(self, *args, **kwargs):
+        """Construct the instance."""
+        super().__init__(*args, **kwargs)
+
+        if self._calculation_class is None or not issubclass(self._calculation_class, CalcJob):
+            raise ValueError('no valid CalcJob class defined for `_calculation_class` attribute')
+
+        self._load_error_handlers()
+
+    @override
+    def load_instance_state(self, saved_state, load_context):
+        """Load the process instance from a saved state.
+
+        :param saved_state: saved state of existing process instance
+        :param load_context: context for loading instance state
+        """
+        super().load_instance_state(saved_state, load_context)
+        self._load_error_handlers()
+
+    def _load_error_handlers(self):
```

Comment:
Shouldn't there also be a ``_load_sanity_checks``? Or if the two are supposed to be in the same entry points list (which I guess they might as well), maybe the ``cls._error_handler_entry_point`` should be renamed to something more generic, like ``cls._handler_entry_points``?

Reply:
Good point, I think they can be in the same entry point, because the decorator will distinguish them. I am a bit annoyed about the asymmetry in `sanity_check` and `error_handler` and that they don't share a common suffix. Because this way `_handler_entry_points` may be mistaken to only reference error handlers.

Reply:
Maybe call it ``_registration_entry_points`` instead?